pageserver/src/consumption_metrics.rs

```diff
@@ -166,14 +169,34 @@ pub async fn collect_metrics_iteration(
             if timeline.is_active() {
                 let timeline_written_size = u64::from(timeline.get_last_record_lsn());
 
-                current_metrics.push((
-                    PageserverConsumptionMetricsKey {
-                        tenant_id,
-                        timeline_id: Some(timeline.timeline_id),
-                        metric: WRITTEN_SIZE,
-                    },
-                    timeline_written_size,
-                ));
+                let key = PageserverConsumptionMetricsKey {
+                    tenant_id,
+                    timeline_id: Some(timeline.timeline_id),
+                    metric: WRITTEN_SIZE,
+                };
+
+                // last_record_lsn can only go up, right now at least, TODO: #2592 or related
+                // features might change this.
+                //
+                // this will be None for the first item, do not send it then to avoid the need to
+                // filter out huge values from the stream of deltas.
+                let timeline_written_size_delta = cached_metrics
+                    .get(&key)
+                    .copied()
+                    .map(|prev| timeline_written_size - prev);
+
+                current_metrics.push((key, timeline_written_size));
+
+                if let Some(timeline_written_size_delta) = timeline_written_size_delta {
```

Comment:
Should you send the same value as you sent in `WRITTEN_SIZE` if None?

Reply:
It would produce:

```
0, 1, 5, 2, 3, <restart>, 11, 1, 2, 3
```

I am not sure if the receiver wants to filter out these large values. I assume they don't.

Reply:
> I am not sure if the receiver wants to filter out these large values. I assume they don't

Yes, that's correct, we don't want jumps between deltas and absolute value for the same metric.

At the same time, if calculation interval is 10 mins, it looks like we will always skip one 10 min interval for this metric after restart:
- [0m, 10m] (after start) - we put value N1 instead of None into cache
- [10m, 20m] - we put N2 into cache and report `N2 - N1` as delta

Right?

If so, it will eat our SLO budget (each metric should arive every 15 mins).

Can we use initialize cache with `last_record_lsn` at start? IIUC, it should be known and stored in the timeline metadata.

Another option is to launch `written_data_delta_bytes` 'calculation' right after pageserver start. That way we put current `last_record_lsn` into cache and in 10 mins after start we report proper deltas.

What do you think?

Reply:
> Can we initialize cache with last_record_lsn at start? IIUC, it should be known and stored in the timeline metadata.
> 
> Another option is to launch written_data_delta_bytes 'calculation' right after pageserver start. That way we put current last_record_lsn into cache and in 10 mins after start we report proper deltas.

Yes, we can easily capture the load time last_record_lsn (disk_consistent_lsn).

However ... Can we be sure this is monotonous? In practice, we rely on safekeepers, and as such, we tolerate losing the latest LSN's under normal operation, which is entirely fine because we'll just restart streaming WAL from an earlier LSN.

Reply:
The solution I ended up with the history up to [2ea0a3a](https://github.com/neondatabase/neon/pull/4822/commits/2ea0a3a7d4ed716485eed96ff0b0722c3cff6bc5) is to always produce a delta, which will likely be 0 for the first round. It might be more, if there was WAL to catch up to from safekeepers.

I am trying to come up with an explanation of how much we could gain error where `error = written_size - sum(written_size_delta_bytes)` in which conditions, but it requires fixing the last time we sent over `written_size` compared to how much WAL the timeline receives between then and the shutdown, so feeling like this is quite difficult to put down as words.

Reply:
Linked commit above will be squashed in rebase, but not much has changed since. I fixed the start time of the event to be either:

1. latest written_size_delta_bytes (incremental)
2. latest written_size (absolute)
3. loaded_at.1 (loading timestamp)